### PR TITLE
New: py-ax-platform; Update: py-botorch, py-gpytorch, py-linear-operator, py-pyro-ppl, py-typeguard

### PR DESCRIPTION
**Pull request by meyersbs**
```mermaid
graph TD;
  'py-ax-platform'-->'py-botorch';
  'py-ax-platform'-->'py-typeguard';
  'py-botorch'-->'py-linear-operator';
  'py-botorch'-->'py-pyro-ppl';
  'py-botorch'-->'py-gpytorch';
```

### CODE REVIEW

**Comment by meyersbs** [comment]
@spackbot fix style

**Comment by meyersbs** [comment]
@spackbot fix style

**Comment by meyersbs** [comment]
Something weird with style checks; first it says something's wrong, then it passes without making changes.

File: var/spack/repos/builtin/packages/py-ax-platform/package.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    maintainers("meyersbs")
+
+    version("0.3.1", sha256="0bad1d16155560fdd8644308d2771edf7fd977ad41fea15a7ecf3f224bc36517")
+
+    # From pyproject.toml
+    depends_on("py-setuptools@34.4:", type="build")
+    depends_on("py-setuptools-scm", type="build")
+    # From setup.py
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("py-botorch@0.8.3:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ax-platform/package.py:26]: I think you need to set the `ALLOW_BOTORCH_LATEST` env var in order to support newer botorch

File: var/spack/repos/builtin/packages/py-ax-platform/package.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    depends_on("py-jinja2", type=("build", "run"))
+    depends_on("py-pandas", type=("build", "run"))
+    depends_on("py-scipy", type=("build", "run"))
+    depends_on("py-scikit-learn", type=("build", "run"))
+    depends_on("py-ipywidgets", type=("build", "run"))
+    depends_on("py-typeguard@2.13.3:", type=("build", "run"))
+    depends_on("py-plotly@5.12.0:", when="^py-ipywidgets@8.0.0:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ax-platform/package.py]:
```suggestion
    depends_on("py-plotly@5.12.0:", type=("build", "run"))
```
Always required

File: var/spack/repos/builtin/packages/py-ax-platform/package.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    depends_on("py-setuptools-scm", type="build")
+    # From setup.py
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("py-botorch@0.8.3:", type=("build", "run"))
+    depends_on("py-jinja2", type=("build", "run"))
+    depends_on("py-pandas", type=("build", "run"))
+    depends_on("py-scipy", type=("build", "run"))
+    depends_on("py-scikit-learn", type=("build", "run"))
+    depends_on("py-ipywidgets", type=("build", "run"))
+    depends_on("py-typeguard@2.13.3:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ax-platform/package.py]:
```suggestion
    depends_on("py-typeguard@2.13.3", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-botorch/package.py [hunk 4]
```diff
@@ -14,15 +14,21 @@ class PyBotorch(PythonPackage):
...
 
+    version("0.8.3", sha256="e529f7adbb2b54f46125ae904682fc0f0d02ab8bdb9067ede521c379b355bf73")
     version("0.6.4", sha256="3fd28417f55749501a45378f72cd5ca7614e2e05b7b65c6b4eb9b72378bc665a")
 
     depends_on("python@3.7:", type=("build", "run"))
-    # TODO: replace this after concretizer learns how to concretize separate build deps
+    depends_on("python@3.8:", when="@0.8.3:", type=("build", "run"))
     depends_on("py-setuptools", type="build")
-    # depends_on('py-setuptools@:47', type='build')
+    depends_on("py-setuptools@:47", when="@:0.6.4", type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-botorch/package.py]:
```suggestion
    depends_on("py-setuptools@:47", when="@:0.6.4", type="build")
```
To fix style (not sure why it didn't work)

File: var/spack/repos/builtin/packages/py-botorch/package.py [hunk 5]
```diff
@@ -14,15 +14,21 @@ class PyBotorch(PythonPackage):
...
+    depends_on("py-torch@1.12:", when="@0.8.3:", type=("build", "run"))
     depends_on("py-torch@1.9:", type=("build", "run"))
+    depends_on("py-gpytorch@1.9.1:", when="@0.8.3:", type=("build", "run"))
     depends_on("py-gpytorch@1.6:", type=("build", "run"))
     depends_on("py-scipy", type=("build", "run"))
     depends_on("py-multipledispatch", type=("build", "run"))
-    depends_on("py-pyro-ppl@1.8.0", type=("build", "run"))
+    depends_on("py-pyro-ppl@1.8.4:", when="@0.8.3:", type=("build", "run"))
+    depends_on("py-pyro-ppl@1.8.0", when="@:0.6.4", type=("build", "run"))
+    depends_on("py-linear-operator@0.3.0:", when="@0.8.3:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-botorch/package.py:37]: Need to set `ALLOW_LATEST_GPYTORCH_LINOP` env var in order to use newer gpytorch/linear-operator

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Glad someone else is using these packages!

**Review by tldahlgren — COMMENTED** [review]
Confirmed the sha256 for all 9 versions.

File: var/spack/repos/builtin/packages/py-ax-platform/package.py (hunk 1 above)
Comment (meyersbs) [reply var/spack/repos/builtin/packages/py-ax-platform/package.py:26]: How should I handle this? A variant to set that flag, defaulting to false?

**Comment by meyersbs** [comment]
@spackbot fix style

**Comment by meyersbs** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/py-ax-platform/package.py (hunk 1 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-ax-platform/package.py:26]: Nope, just always set it

File: var/spack/repos/builtin/packages/py-plotly/package.py [hunk 6]
```diff
@@ -12,13 +12,17 @@ class PyPlotly(PythonPackage):
...
 
+    maintainers("meyersbs")
+
+    version("5.14.1", sha256="bcac86d7fcba3eff7260c1eddc36ca34dae2aded10a0709808446565e0e53b93")
     version("5.2.2", sha256="809f0674a7991daaf4f287964d617d24e9fa44463acd5a5352ebd874cfd98b07")
     version("2.2.0", sha256="ca668911ffb4d11fed6d7fbb12236f8ecc6a7209db192326bcb64bdb41451a58")
 
     depends_on("python@3.6:", when="@5.2.2:", type=("build", "run"))
 
     depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-plotly/package.py:23]: Missing build dep on jupyterlab

File: var/spack/repos/builtin/packages/py-plotly/package.py [hunk 7]
```diff
@@ -12,13 +12,17 @@ class PyPlotly(PythonPackage):
...
-    depends_on("py-six", type=("build", "run"))
+    depends_on("py-setuptools@40.8.0:", when="@5.14.1:", type="build")
+    depends_on("py-six", when="@:5.2.2", type=("build", "run"))
 
     depends_on("py-pytz", when="@:2.2.0", type=("build", "run"))
     depends_on("py-decorator@4.0.6:", when="@:2.2.0", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-plotly/package.py:29]: Missing build/run dep on packaging

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-plotly/package.py [hunk 8]
```diff
@@ -12,17 +12,23 @@ class PyPlotly(PythonPackage):
...
+    version("5.14.1", sha256="bcac86d7fcba3eff7260c1eddc36ca34dae2aded10a0709808446565e0e53b93")
     version("5.2.2", sha256="809f0674a7991daaf4f287964d617d24e9fa44463acd5a5352ebd874cfd98b07")
     version("2.2.0", sha256="ca668911ffb4d11fed6d7fbb12236f8ecc6a7209db192326bcb64bdb41451a58")
 
     depends_on("python@3.6:", when="@5.2.2:", type=("build", "run"))
 
     depends_on("py-setuptools", type="build")
-    depends_on("py-six", type=("build", "run"))
+    depends_on("py-setuptools@40.8.0:", when="@5.14.1:", type="build")
+    depends_on("py-jupyterlab@:2.9,3.1:", when="@5.14.1", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-plotly/package.py]: `~=` means the exact opposite of what you think it means:
```suggestion
    depends_on("py-jupyterlab@3", when="@5:", type="build")
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Comment by meyersbs** [comment]
Apparently Monday brain has carried over into Tuesday... Thanks for all of your feedback!

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**